Release/src/http/listener/http_server_named_pipe.cpp

```diff
@@ -0,0 +1,728 @@
+    {
+        size_t bytes_read = 0;
+
+        // If an exception occurs surface the error to user on the server side
+        // and cancel the request so the client sees the error.
+        try { 
+            bytes_read = op.get(); 
+        }
+        catch (...)
+        {
+            cancel_request(std::current_exception());
+        }
+        if (bytes_read == 0)
+        {
+            cancel_request(std::make_exception_ptr(http_exception(_XPLATSTR("Error unexpectedly encountered the end of the response stream early"))));
+        }
+        return bytes_read;
+    }).get();
+
+    for (size_t i = 0; i < bytes_read; i++)
+        os << static_cast<char>(m_body_data[i]);
+
+    m_overlapped.set_http_io_completion([this](DWORD error, DWORD nBytes) {
+        on_response_written(error, nBytes);
+    });
+
+    StartThreadpoolIo(m_threadpool_io);
+
+    DWORD bytesWritten = 0;
+
+    auto success = WriteFile(m_pipeHandle, os.str().c_str(), static_cast<DWORD>(os.str().length()), &bytesWritten, &m_overlapped);
+
+    if (!success && GetLastError() != ERROR_IO_PENDING)
+    {
+        CancelThreadpoolIo(m_threadpool_io);
+        m_msg.reply(status_codes::InternalError);
+        init_response_callbacks(ShouldWaitForBody::DontWait);
+    }
+}
+
+/// <summary>
+///  The send response body completion callback function.
+/// </summary>
+void named_pipe_request_context::on_response_written(DWORD error_code, DWORD)
+{
+    if (error_code != NO_ERROR)
+    {
+        cancel_request(std::make_exception_ptr(http_exception(error_code)));
+    }
+    else
+    {
+        std::unique_lock<std::mutex> lock(m_responseCompletedLock);
+        m_response_completed.set();
+    }
+}
+
+void named_pipe_request_context::cancel_request(std::exception_ptr except_ptr)
+{
+    // TODO 
+}
+
+/// <summary>
+///  The cancel request completion callback function.
+/// </summary>
+void named_pipe_request_context::on_request_canceled(DWORD, DWORD)
+{
+    std::unique_lock<std::mutex> lock(m_responseCompletedLock);
+    m_response_completed.set_exception(m_except_ptr);
+}
+
+
+/***************************************************************************************************
+
+                                named_pipe_listener
+
+***************************************************************************************************/
+
+named_pipe_listener::named_pipe_listener(web::http::experimental::listener::details::http_listener_impl* listener)
+    : m_listener(listener)
+{}
+
+void named_pipe_listener::start() {
+    async_receive_request();
+}
+
+void named_pipe_listener::async_receive_request()
+{
+    // TODO: validate uri
+    auto path = uri::split_path(m_listener->uri().path());
+
+    utility::string_t pipe_name = U("\\\\.\\pipe\\") + path[0];
+
+    //
+    // We use the following ACL for the named pipe:
+    //
+    auto acl = L"D:(A;;GA;;;BA)";
```

Comment:
>acl = L"D:(A;;GA;;;BA)"; [](start = 9, length = 24)

the ACL seems right, however this link mentions to explicitly deny access to NT Authority\network service [https://docs.microsoft.com/en-us/windows/desktop/ipc/named-pipes](https://docs.microsoft.com/en-us/windows/desktop/ipc/named-pipes) . If we don't need to explicitly disable it than we are good.

Reply:
Yup, added DENY, thanks!